harmonica/tests/test_legendre.py

```diff
@@ -0,0 +1,203 @@
+    associated_legendre_schmidt_derivative,
+)
+
+
+def legendre_analytical(x):
+    "Analytical expressions for unnormalized Legendre functions"
+    max_degree = 4
+    p = np.full((max_degree + 1, max_degree + 1), np.nan)
+    p[0, 0] = 1
+    p[1, 0] = x
+    p[1, 1] = np.sqrt(1 - x**2)
+    p[2, 0] = 1 / 2 * (3 * x**2 - 1)
+    p[2, 1] = 3 * x * np.sqrt(1 - x**2)
+    p[2, 2] = 3 * (1 - x**2)
+    p[3, 0] = 1 / 2 * (5 * x**3 - 3 * x)
+    p[3, 1] = -3 / 2 * (1 - 5 * x**2) * np.sqrt(1 - x**2)
+    p[3, 2] = 15 * x * (1 - x**2)
+    p[3, 3] = 15 * np.sqrt(1 - x**2) ** 3
+    p[4, 0] = 1 / 8 * (35 * x**4 - 30 * x**2 + 3)
+    p[4, 1] = 5 / 2 * (7 * x**3 - 3 * x) * np.sqrt(1 - x**2)
+    p[4, 2] = 15 / 2 * (7 * x**2 - 1) * (1 - x**2)
+    p[4, 3] = 105 * x * np.sqrt(1 - x**2) ** 3
+    p[4, 4] = 105 * np.sqrt(1 - x**2) ** 4
+    return p
+
+
+def legendre_derivativeative_analytical(x):
+    """
+    Analytical expressions for theta derivatives of unnormalized Legendre
+    functions
+    """
+    max_degree = 4
+    dp = np.full((max_degree + 1, max_degree + 1), np.nan)
+    cos = x
+    sin = np.sqrt(1 - x**2)
+    dp[0, 0] = 0
+    dp[1, 0] = -sin
+    dp[1, 1] = cos
+    dp[2, 0] = -3 * cos * sin
+    dp[2, 1] = 3 * (cos**2 - sin**2)
+    dp[2, 2] = 6 * sin * cos
+    dp[3, 0] = (3 * sin - 15 * cos**2 * sin) / 2
+    dp[3, 1] = 3 / 2 * (5 * cos**3 - 10 * cos * sin**2 - cos)
+    dp[3, 2] = 30 * cos**2 * sin - 15 * sin**3
+    dp[3, 3] = 45 * sin**2 * cos
+    dp[4, 0] = (15 * cos - 35 * cos**3) * sin / 2
+    dp[4, 1] = (
+        35 * cos**4 - 15 * cos**2 + 15 * sin**2 - 105 * cos**2 * sin**2
+    ) / 2
+    dp[4, 2] = 105 * sin * cos**3 - 105 * cos * sin**3 - 15 * sin * cos
+    dp[4, 3] = 315 * cos**2 * sin**2 - 105 * sin**4
+    dp[4, 4] = 420 * sin**3 * cos
+    return dp
+
+
+def schmidt_normalization(p):
+    "Calculate the Schmidt normalization factor"
+    max_degree = p.shape[0] - 1
+    for n in range(max_degree + 1):
+        for m in range(n + 1):
+            if m > 0:
+                p[n, m] *= np.sqrt(2 * math.factorial(n - m) / math.factorial(n + m))
+
+
+def full_normalization(p):
+    "Calculate the full normalization factor"
+    max_degree = p.shape[0] - 1
+    for n in range(max_degree + 1):
+        for m in range(n + 1):
+            if m == 0:
+                p[n, m] *= np.sqrt(2 * n + 1)
+            else:
+                p[n, m] *= np.sqrt(
+                    2 * (2 * n + 1) * math.factorial(n - m) / math.factorial(n + m)
+                )
+
+
+@pytest.mark.parametrize(
+    "func,norm",
+    (
+        (associated_legendre, None),
+        (associated_legendre_schmidt, schmidt_normalization),
+        (associated_legendre_full, full_normalization),
+    ),
+    ids=["unnormalized", "schmidt", "full"],
+)
+def test_associated_legendre_function_analytical(func, norm):
+    "Check if the first few degrees match analytical expressions"
+    for angle in np.linspace(0, np.pi, 360):
```

Comment:
Just double checking that you actually meant to test with angles in $[0, \pi]$. I think it makes sense: if we extend it to $[0, 2\pi]$ we'll be duplicating the same tests, but just wanted to make sure of it.

Reply:
Yep, that range is the valid range of colatitude, which is the angle used in spherical harmonics.